catalogd/internal/storage/localdir.go

```diff
@@ -20,95 +26,266 @@ import (
+}
+
+func catalogFilePath(catalogDir string) string {
+	return filepath.Join(catalogDir, "catalog.jsonl")
+}
+
+func catalogIndexFilePath(catalogDir string) string {
+	return filepath.Join(catalogDir, "index.json")
 }
 
-// Open opens a named file from the underlying filesystem. If the file
-// is not a regular file, it return os.ErrNotExists. Callers are resposible
-// for closing the file returned.
-func (f *filesOnlyFilesystem) Open(name string) (fs.File, error) {
-	file, err := f.FS.Open(name)
+type storeMetasFunc func(catalogDir string, metaChan <-chan *declcfg.Meta) error
+
+func storeCatalogData(catalogDir string, metas <-chan *declcfg.Meta) error {
+	f, err := os.Create(catalogFilePath(catalogDir))
 	if err != nil {
-		return nil, err
+		return err
+	}
+	defer f.Close()
+
+	for m := range metas {
+		if _, err := f.Write(m.Blob); err != nil {
+			return err
+		}
 	}
-	stat, err := file.Stat()
+	return nil
+}
+
+func storeIndexData(catalogDir string, metas <-chan *declcfg.Meta) error {
+	idx, err := newIndex(metas)
 	if err != nil {
-		_ = file.Close()
-		return nil, err
+		return err
 	}
-	if !stat.Mode().IsRegular() {
-		_ = file.Close()
-		return nil, os.ErrNotExist
+
+	f, err := os.Create(catalogIndexFilePath(catalogDir))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	enc := json.NewEncoder(f)
+	enc.SetEscapeHTML(false)
+	return enc.Encode(idx)
+}
+
+func (s *LocalDirV1) BaseURL(catalog string) string {
+	return s.RootURL.JoinPath(catalog).String()
+}
+
+func (s *LocalDirV1) StorageServerHandler() http.Handler {
+	mux := http.NewServeMux()
+
+	mux.HandleFunc(s.RootURL.JoinPath("{catalog}", "api", "v1", "all").Path, s.handleV1All)
+	if s.EnableQueryHandler {
+		mux.HandleFunc(s.RootURL.JoinPath("{catalog}", "api", "v1", "query").Path, s.handleV1Query)
+	}
+	return mux
+}
+
+func (s *LocalDirV1) handleV1All(w http.ResponseWriter, r *http.Request) {
+	s.m.RLock()
+	defer s.m.RUnlock()
+
+	catalog := r.PathValue("catalog")
+	catalogFile, catalogStat, err := s.catalogData(catalog)
+	if err != nil {
+		httpError(w, err)
+		return
+	}
+	serveJsonLines(w, r, catalogStat.ModTime(), catalogFile)
+}
+
+func (s *LocalDirV1) handleV1Query(w http.ResponseWriter, r *http.Request) {
+	s.m.RLock()
+	defer s.m.RUnlock()
+
+	catalog := r.PathValue("catalog")
+	catalogFile, catalogStat, err := s.catalogData(catalog)
+	if err != nil {
+		httpError(w, err)
+		return
+	}
+	defer catalogFile.Close()
+
+	schema := r.URL.Query().Get("schema")
+	pkg := r.URL.Query().Get("package")
+	name := r.URL.Query().Get("name")
+
+	if schema == "" && pkg == "" && name == "" {
+		// If no parameters are provided, return the entire catalog (this is the same as /api/v1/all)
+		serveJsonLines(w, r, catalogStat.ModTime(), catalogFile)
+		return
+	}
+	idx, err := s.getIndex(catalog)
+	if err != nil {
+		httpError(w, err)
+		return
+	}
+	indexReader, ok := idx.Get(catalogFile, schema, pkg, name)
+	if !ok {
+		httpError(w, fs.ErrNotExist)
+		return
```

Comment:
This should change to write a 200 status and write no content to the response body.

Reply:
I noticed that `index.Get()` was only ever returning `true` (so this code path was never actually being hit). Something else solved the test case failure though, but I went ahead and cleaned up  `Get()`'s signature https://github.com/operator-framework/operator-controller/pull/1643/commits/53692a97707d315c84ce3b592bc77dcf5fc04003